File: clojure/kindergarten-garden/src/kindergarten_garden.clj
(ns kindergarten-garden
  (:require [clojure.string :as str]))

(def students ["alice" "bob" "charlie" "david" "eve" "fred" "ginny" "harriet" "ileana" "joseph" "kincaid" "larry"])

(defn char->flower [c]
  (case c
    \R :radishes
    \C :clover
    \G :grass
    \V :violets))

(defn string->flowers [string]
  (map #(map char->flower %) (str/split-lines string)))

(defn students->keys [students]
  (sort (map (comp keyword str/lower-case) students)))

(defn garden
  ([string] (garden string students))
  ([string students] (->> (string->flowers string)
                          (map #(partition 2 %))
                          (apply map concat)
                          (zipmap (students->keys students)))))

File: ruby/kindergarten-garden/kindergarten_garden.rb
class Garden
  def method_missing(name, *args, &block)
    index = students.index(name.to_s)
    super if index.nil?

    define_singleton_method(name) do
      flowers[index]
    end

    send(name)
  end

  def respond_to_missing?(method_name, include_private = false)
    !students.index(name.to_s).nil? || super
  end

  private

  attr_reader :flowers, :students

  def initialize(diagram, students = %w[alice bob charlie david eve fred ginny harriet ileana joseph kincaid larry])
    @flowers = parse_diagram(diagram)
    @students = students.map(&:downcase).sort
  end

  def parse_diagram(diagram)
    flowers = parse_flowers(diagram)
    line1, line2 = flowers
    line1.each_slice(2).zip(line2.each_slice(2)).map { _1 + _2 }
  end

  def parse_flowers(diagram)
    diagram.split.map do |line|
      line.chars.map do |char|
        flower(char)
      end
    end
  end

  def flower(char)
    {
      'R' => :radishes,
      'C' => :clover,
      'G' => :grass,
      'V' => :violets
    }[char]
  end
end
